Mathlib/Algebra/Hom/Ring.lean

```diff
@@ -0,0 +1,788 @@
+def comp (g : β →+* γ) (f : α →+* β) : α →+* γ :=
+  { g.toNonUnitalRingHom.comp f.toNonUnitalRingHom with toFun := g ∘ f, map_one' := by simp }
+#align ring_hom.comp RingHom.comp
+
+/-- Composition of semiring homomorphisms is associative. -/
+theorem comp_assoc {δ} {_ : NonAssocSemiring δ} (f : α →+* β) (g : β →+* γ) (h : γ →+* δ) :
+    (h.comp g).comp f = h.comp (g.comp f) :=
+  rfl
+#align ring_hom.comp_assoc RingHom.comp_assoc
+
+@[simp]
+theorem coe_comp (hnp : β →+* γ) (hmn : α →+* β) : (hnp.comp hmn : α → γ) = hnp ∘ hmn :=
+  rfl
+#align ring_hom.coe_comp RingHom.coe_comp
+
+theorem comp_apply (hnp : β →+* γ) (hmn : α →+* β) (x : α) :
+    (hnp.comp hmn : α → γ) x = hnp (hmn x) :=
+  rfl
+#align ring_hom.comp_apply RingHom.comp_apply
+
+@[simp]
+theorem comp_id (f : α →+* β) : f.comp (id α) = f :=
+  ext fun _ => rfl
+#align ring_hom.comp_id RingHom.comp_id
+
+@[simp]
+theorem id_comp (f : α →+* β) : (id β).comp f = f :=
+  ext fun _ => rfl
+#align ring_hom.id_comp RingHom.id_comp
+
+instance : Monoid (α →+* α) where
+  one := id α
+  mul := comp
+  mul_one := comp_id
+  one_mul := id_comp
+  mul_assoc f g h := comp_assoc _ _ _
+
+theorem one_def : (1 : α →+* α) = id α :=
+  rfl
+#align ring_hom.one_def RingHom.one_def
+
+theorem mul_def (f g : α →+* α) : f * g = f.comp g :=
+  rfl
+#align ring_hom.mul_def RingHom.mul_def
+
+@[simp]
+theorem coe_one : ⇑(1 : α →+* α) = _root_.id :=
+  rfl
+#align ring_hom.coe_one RingHom.coe_one
+
+@[simp]
+theorem coe_mul (f g : α →+* α) : ⇑(f * g) = f ∘ g :=
+  rfl
+#align ring_hom.coe_mul RingHom.coe_mul
+
+theorem cancel_right {g₁ g₂ : β →+* γ} {f : α →+* β} (hf : Surjective f) :
+    g₁.comp f = g₂.comp f ↔ g₁ = g₂ :=
+  ⟨fun h => RingHom.ext <| hf.forall.2 (ext_iff.1 h), fun h => h ▸ rfl⟩
+#align ring_hom.cancel_right RingHom.cancel_right
+
+theorem cancel_left {g : β →+* γ} {f₁ f₂ : α →+* β} (hg : Injective g) :
+    g.comp f₁ = g.comp f₂ ↔ f₁ = f₂ :=
+  ⟨fun h => RingHom.ext fun x => hg <| by rw [← comp_apply, h, comp_apply], fun h => h ▸ rfl⟩
+#align ring_hom.cancel_left RingHom.cancel_left
+
+end RingHom
+
+/-- Pullback `IsDomain` instance along an injective function. -/
+protected theorem Function.Injective.isDomain [Ring α] [IsDomain α] [Ring β] (f : β →+* α)
+    (hf : Injective f) : IsDomain β := by
+  haveI := pullback_nonzero f f.map_zero f.map_one
+  haveI := IsRightCancelMulZero.toNoZeroDivisors α
+  haveI := hf.noZeroDivisors f f.map_zero f.map_mul
+  exact NoZeroDivisors.toIsDomain β
+#align function.injective.is_domain Function.Injective.isDomain
+
+namespace AddMonoidHom
+
+variable [CommRing α] [IsDomain α] [CommRing β] (f : β →+ α)
+
+/-- Make a ring homomorphism from an additive group homomorphism from a commutative ring to an
+integral domain that commutes with self multiplication, assumes that two is nonzero and `1` is sent
+to `1`. -/
+def mkRingHomOfMulSelfOfTwoNeZero (h : ∀ x, f (x * x) = f x * f x) (h_two : (2 : α) ≠ 0)
```

Comment:
This naming surprised me a bit; does the `_of_` naming scheme not apply for data?

Reply:
I'm not aware of any exceptions to data naming scheme when `of` is involved, and I would slightly prefer not creating such an exception

Reply:
Asked on zulip: https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/naming.20conventions/near/315558410